### PR TITLE
Reimagine Sharing: Add new sharing UI with shareable image preview

### PR DESCRIPTION
**Pull request by bjtitus**
| 📘 Part of: https://github.com/Automattic/pocket-casts-ios/issues/1910 |
|:---:|

This adds the beginning of the new UI for sharing Episodes, Podcasts, and Current Position with an image asset.

| Episode | Podcast | Current Position |
| -- | -- | -- |
| Simulator Screenshot - iPhone 15 Pro - 2024-06-20 at 14 22 40 | Simulator Screenshot - iPhone 15 Pro - 2024-06-20 at 14 22 46 | Simulator Screenshot - iPhone 15 Pro - 2024-06-20 at 14 22 51 |

We will likely need to tweak a few small design things, like the gradient used for the shared asset, but this can be done in a follow up PR.

A few things are missing from this first PR:
* Sharing options - to keep the line count low for review
* Sharing from the Episode page - still working on a few edge cases there
* Asset generation + sharing - need sharing actions first

## To test

> [!NOTE]
> Enable the `newSharing` feature flag to test this

### Test Sharing from Podcast page
* Visit a Podcast page
* Tap the Share icon at the top
* Verify the new sharing appears

### Test Sharing from Now Playing
* Play a Podcast episode
* Share from the Now Playing Screen
* Choose Episode, Current Position, or Podcast
* Verify that the image that shows appears correct

## Test different image sizes
* Swiping left and right should display differently sized share images. Make sure these appear properly.

## Checklist

- [x] I have considered if this change warrants user-facing release notes and have added them to `CHANGELOG.md` if necessary.
- [x] I have considered adding unit tests for my changes.
- [x] I have updated (or requested that someone edit) [the spreadsheet](https://docs.google.com/spreadsheets/d/107jqrutZhU0fVZJ19SBqxxVKbV2NWSdQC9MFYdLiAxc/edit?usp=sharing) to reflect any new or changed analytics.

### CODE REVIEW

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/Automattic/pocket-casts-ios/milestone/53">7.67</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Review by leandroalonso — CHANGES_REQUESTED** [review]
@bjtitus I'm facing two different issues:

1. When sharing from the player, the `assertionFailure` is failing. However by using `.parentPodcast()` I see it's there:

Screenshot 2024-06-20 at 17 20 50

2. Inside the episode screen (regarding the testing steps at Test Sharing from Episode page), the share is the same as before (not sure if that's missing due to splitting the PRs):

https://github.com/Automattic/pocket-casts-ios/assets/7040243/145f94e4-f2c2-44c6-99cc-d1f631b1181c

File: podcasts/NowPlayingPlayerItemViewController+Shelf.swift [hunk 1]
```diff
@@ -414,9 +414,9 @@ extension NowPlayingPlayerItemViewController: NowPlayingActionsDelegate {
 
         if FeatureFlag.newSharing.enabled {
             if fromTime == 0 {
-                SharingModal.show(option: .episode, episode: episode, in: self)
+                SharingModal.show(option: .episode(episode), episode: episode, in: self)
```
Comment (leandroalonso) [inline podcasts/NowPlayingPlayerItemViewController+Shelf.swift]: Do we really need to give the same `episode` argument twice?

File: podcasts/NowPlayingPlayerItemViewController+Shelf.swift [hunk 2]
```diff
@@ -429,7 +429,7 @@ extension NowPlayingPlayerItemViewController: NowPlayingActionsDelegate {
 
         Analytics.track(.podcastShared, properties: ["type": "podcast", "source": "player"])
         if FeatureFlag.newSharing.enabled {
-            SharingModal.show(option: .episode, podcast: podcast, episode: nil, in: self)
+            SharingModal.show(option: .podcast(podcast), podcast: podcast, episode: nil, in: self)
```
Comment (leandroalonso) [inline podcasts/NowPlayingPlayerItemViewController+Shelf.swift]: Same question as before: can't we just give `podcast` once?

File: podcasts/Sharing/PocketCastsLogoPill.swift [hunk 3]
```diff
@@ -0,0 +1,30 @@
+//
+//  PocketCastsLogoPill.swift
+//  podcasts
+//
+//  Created by Brandon Titus on 4/17/24.
+//  Copyright © 2024 Shifty Jelly. All rights reserved.
+//
+
```
Comment (leandroalonso) [inline podcasts/Sharing/PocketCastsLogoPill.swift]:
```suggestion
```

**Comment by bjtitus** [comment]
> When sharing from the player, the assertionFailure is failing. However by using .parentPodcast() I see it's there

Interesting, I hadn't seen this yet but I'll just switch to `parentPodcast()` here. Maybe this is because `parentPodcast` also searches unsubscribed podcasts?

> the share is the same as before

Yes, that's expected. Sorry, I've removed that testing step. There are a few more menu options that I need to account for.

**Comment by leandroalonso** [comment]
> Maybe this is because parentPodcast also searches unsubscribed podcasts?

That's probably it. I wasn't subscribed to this podcast

File: podcasts/Sharing/ShareImageView.swift [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+            Text(info.description)
+                .font(.caption)
+                .fontWeight(.semibold)
+                .foregroundStyle(.white.opacity(0.5))
+        }
+        .multilineTextAlignment(textAlignment)
+    }
+}
+
+let previewInfo = ShareImageInfo(name: "This American Life", title: "Dylan Field, Figma Co-founder, Talks Design, Economy, and life after failed Adobe acquisitions", description: Date().formatted(), artwork: URL(string: "")!, gradient: Gradient(colors: [Color.red, Color(hex: "620603")]))
```
Comment (leandroalonso) [inline podcasts/Sharing/ShareImageView.swift]: The `URL(string: "")!` is failing which makes the preview unavailable. Adding any valid URL fixes it.

File: podcasts/Sharing/ShareImageView.swift [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+                .padding(24)
+                .frame(width: 324, height: 169)
+            }
+        }
+        .fixedSize()
+        .clipShape(RoundedRectangle(cornerRadius: 12))
+    }
+
+    @ViewBuilder func image() -> some View {
+        KFImage(info.artwork)
```
Comment (leandroalonso) [inline podcasts/Sharing/ShareImageView.swift:72]: I believe we should use episode artwork if available (and enabled). Not necessarily needs to be addressed here (we can follow-up on a subsequent PR)

File: podcasts.xcodeproj/project.pbxproj [hunk 6]
```diff
@@ -11219,7 +11235,10 @@
 			buildSettings = {
 				ASSETCATALOG_COMPILER_APPICON_NAME = AppIconPrototype;
 				CLANG_ENABLE_MODULES = YES;
-				CODE_SIGN_ENTITLEMENTS = podcasts/podcasts.prototype.entitlements;
+				CODE_SIGN_ENTITLEMENTS = podcasts/podcastsDebug.entitlements;
```
Comment (leandroalonso) [inline podcasts.xcodeproj/project.pbxproj]: Just wondering if these changes are Xcode stuff or something else?

**Review by leandroalonso — CHANGES_REQUESTED** [review]
Tested and it works as described. Left a few comments to improve code and some questions!

File: podcasts.xcodeproj/project.pbxproj (hunk 6 above)
Comment (bjtitus) [reply podcasts.xcodeproj/project.pbxproj]: I _think_ they were Xcode but I went in and undid them and Xcode hasn't added them back yet.

File: podcasts/Sharing/ShareImageView.swift (hunk 5 above)
Comment (bjtitus) [reply podcasts/Sharing/ShareImageView.swift:72]: 👍 Good catch! I've added this as a task to the project thread so I don't forget this.

**Review by leandroalonso — APPROVED** [review]
